dap4/d4tests/src/test/java/dap4/test/TestDAP4Client.java

```diff
@@ -0,0 +1,406 @@
+      if (constraint == null)
+        constraint = "";
+      // Break off the final file set name
+      int index = datasetpath.lastIndexOf('/');
+      this.dataset = datasetpath.substring(index + 1, datasetpath.length());
+      this.title = this.dataset;
+      this.id = id;
+      this.constraint = (constraint.length() == 0 ? null : constraint);
+      this.baselinepath = root + "/" + BASELINEDIR + "/" + dataset;
+      if (this.constraint != null)
+        this.baselinepath += ("." + String.valueOf(this.id));
+    }
+
+    public ClientTest nochecksum() {
+      this.checksumming = false;
+      return this;
+    }
+
+    public ClientTest xfail() {
+      this.xfail = true;
+      return this;
+    }
+
+    public ClientTest headeronly() {
+      this.headeronly = true;
+      return this;
+    }
+
+    String makeurl() {
+      String url = this.server + "/" + this.servlet + "/" + this.dataset;
+      if (constraint != null)
+        url += ("?" + constraint);
+      url += "#dap4";
+      return url;
+    }
+
+    public String toString() {
+      return dataset;
+    }
+  }
+
+  //////////////////////////////////////////////////
+  // Class variables
+
+  // Order is important; testing reachability is in the order listed
+  static List<Server> SERVERS;
+
+  static {
+    SERVERS = new ArrayList<>();
+    SERVERS.add(new Server("149.165.169.123", "8080"));
+    SERVERS.add(new Server("remotetest.unidata.ucar.edu", null));
+  }
+
+  //////////////////////////////////////////////////
+  // Instance variables
+
+  // Test cases
+
+  List<ClientTest> alltestcases = new ArrayList<ClientTest>();
+  List<ClientTest> chosentests = new ArrayList<ClientTest>();
+
+  String datasetpath = null;
+
+  Server server = null;
+
+  //////////////////////////////////////////////////
+
+  @Before
+  public void setup() throws Exception {
+    // Find the server to use
+    this.server = null;
+    for (Server svc : SERVERS) {
+      String url = svc.getTestURL();
+      try (HTTPMethod method = HTTPFactory.Get(url)) {
+        try {
+          int code = method.execute();
+          if (code == 200) {
+            this.server = svc;
+            System.out.println("Using server url " + url);
+            break;
+          }
+        } catch (HTTPException e) {
+          this.server = null;
+        }
+      }
+    }
+    if (this.server == null)
+      throw new Exception("Cannot locate server");
+    defineAllTestcases(this.getResourceDir(), SERVLETPATH, this.server.getURL());
+    chooseTestcases();
+    if (BUILDBASELINE)
+      prop_baseline = true;
+  }
+
+  //////////////////////////////////////////////////
+  // Define test cases
+
+  void chooseTestcases() {
+    if (false) {
+      chosentests = locate("test_atomic_types.nc");
+      prop_baseline = true;
+    } else {
+      for (ClientTest tc : alltestcases) {
+        chosentests.add(tc);
+      }
+    }
+  }
+
+  boolean defineAllTestcases(String root, String servlet, String server) {
```

Comment:
JUnit parameterized?

Reply:
I'm happy to take a pass at converting this test to use parameterized if you'd like

Reply:
I would say don't bother with fixing DAP4 tests at the moment.
This PR addresses issues completely independent of DAP4.
I do not think there is any point in trying to add DAP4 fixes to it
since I have another branch that is working on that. In any case,
converting the DAP4  tests to use parameterized tests is going to take
time. I think we will have to live with it for a while.